### ISSUE TITLE
Feature: support array nodes in schema

### ISSUE THREAD

**Issue by pedro-lb** [jungsoft/formup#32]
# Description

We need to support array nodes in yup schemas, anything from a simple array of strings to a complex array of objects.

Any array would produce multiple items in the form (multiple inputs, or multiple input groups with many inputs inside them). We need to be able to access the exact path on the array in order to update the exact form value path.

# Use-case examples

## Array of primitive types / strings

This schema:

```js
const schema = yup.object().shape({
  emergencyPhoneNumbers: yup
    .array()
    .of(yup.string())
    .label("Emergency numbers"),
});
```

Would result in something like: 

![image](https://user-images.githubusercontent.com/22061051/87694210-46667c00-c764-11ea-877c-774b077e95dd.png)

*Note:* interface is not a formup problem, would be implemented by the user. This is just an example.

## Array of objects

This schema:

```js
const personSchema = yup.object().shape({
  name: yup
    .string()
    .required()
    .label("Name"),
  age: yup
    .number()
    .label("Age"),
  email: yup
    .string()
    .required()
    .label("Email"),
});

const schema = yup.object().shape({
  people: yup
    .array()
    .of(personSchema)
    .label("People"),
});
```
Would result in something like: 

![image](https://user-images.githubusercontent.com/22061051/87695009-4dda5500-c765-11ea-8045-ca6ca76b2bb5.png)

*Note:* interface is not a formup problem, would be implemented by the user. This is just an example.

# Implementation use-case example

We could implement three new items that would be returned by `useFormup`:
- *FormArrayField*: A component to help render & build array items;
- *addArrayFieldItem*: A function to add a new array field easily (examples below);
- *removeArrayFieldItem*: A function to remove array fields easily (examples below);

Then, for the examples above, we could implement them this way:

## Array of primitive types / strings

```jsx
const Example = () => {
  const {
    FormInput,
    FormArrayField,
    removeArrayFieldItem,
    addArrayFieldItem,
  } = useFormup(schema);
  
  return (
    <div>
      <FormArrayField name="phones">
        {(items) => items.map((item, index) => (
          <>
            <FormInput
              name={item.path} // item.path (calculated by formup) -> `phones[${index}]`
              // value={item.value.quantity} // we can also access the value at the index in "item"
            />
            
            <Button
              onClick={() => removeArrayFieldItem("phones", index)}
              text="Remove this phone"
            />
          </>
        ))}
      </FormArrayField>
  
      <Button
        onClick={() => addArrayFieldItem("phones")}
        text="Add empty phone"
      />
  
      <Button
        onClick={() => addArrayFieldItem("phones", "109123801234")}
        text="Add new phone"
      />
    </div>
  );
};
```

## Array of objects

```jsx
const Example = () => {
  const {
    FormInput,
    FormArrayField,
    removeArrayFieldItem,
    addArrayFieldItem,
  } = useFormup(schema);
  
  return (
    <div>
      <FormArrayField name="people">
        {(items) => items.map((item, index) => (
          <>
            <FormInput
              name={item.getPath("name")} // here we're accessing "name" inside the array. Full path = people[index].name
              // value={item.value.name} // we can also access the value at the index in "item"
            />

            <FormInput
              name={item.getPath("age")} // here we're accessing "age" inside the array. Full path = people[index].age
              // value={item.value.age} // we can also access the value at the index in "item"
            />

            <FormInput
              name={item.getPath("email")} // here we're accessing "email" inside the array. Full path = people[index].email
              // value={item.value.email} // we can also access the value at the index in "item"
            />
  
            <Button
              onClick={() => removeArrayFieldItem("people", index)}
              text="Remove this person"
            />
          </>
        ))}
      </FormArrayField>
  
      <Button
        onClick={() => removeArrayFieldItem("people", 3)}
        text="Remove person at index 3"
      />
  
      <Button
        onClick={() => addArrayFieldItem("people")}
        text="Add empty person"
      />
  
      <Button
        onClick={() => {
          addArrayFieldItem("people", {
            name: "John Foo",
            age: 10,
            email: "john@foo.com"
          });
        }}
        text="Add new person"
      />
    </div>
  )
};
```

# Implementation details

### Formup needs to be able to understand array positions in `name` properties, in `FormInput`.

For example:

- For primitive types / strings: 

```jsx
<FormInput
  name="phoneNumbers[3]" // on change, would update the item at index 3
/>
```

- For objects:

```js
<FormInput
  name="people[3].name" // on change, would update the property "name" of the item at index 3
/>
```

### New component: `FormArrayField`

It needs to receive a component or function on `children` and pass the required props to it in order to render.

It also needs to throw an exception in case the `name` passed is not an array in schema.

#### Interface (props to receive)

```jsx
interface FormArrayField {
  /**
   * The field name on schema.
   */ 
  name: string;
}
```

#### Interface (props to inject into rendered component)

```jsx
interface FormArrayFieldItem {
  /**
   * The calculated path, so the user doesn't need to string interpolate.
   * Example: phoneNumbers[3]
   */ 
  path: string;

  /**
   * A function to build the path, when using object arrays,
   * so the user doesn't need to string interpolate.
   * Example: getPath("email") -> people[3].email
   */ 
  getPath: (value: string) => string;

  /**
   * The value of the field on formik form.
   */
  value: string;
}

// This would be injected into the rendered component
interface FormArrayFieldInjectProps {
  items: Array<FormArrayFieldItem>;
}
```

#### Render strategy:
  - Render function with arguments if `children` is a function
  - Render component with props if `children` is a component

```jsx
const RenderExample = ({
  children
}) => {
  const props = {
    x: 10,
  },

  if (isFunction(children)) {
    return children(props);
  }

  return React.cloneElement(children, props);
};
```

So that we can support these two cases:

```jsx
// Children as function
const Component = () => (
  <RenderExample>
    {(props) => (
      <div x={props.x}>
        <input />
      </div>
    )}
  </RenderExample>
);
```

```jsx
// Children as component
const InnerComponent = (props) => (
  <div x={props.x}>
    <input />
  </div>
)

const Component = () => (
  <RenderExample>
    <InnerComponent />
  </RenderExample>
)
```

#### Identity

Should add an array item in given schema field.

Should by default use the default value for that field on the schema, but this can be overwritten by the second argument to initialize the value.

```jsx
const addArrayFieldItem = (
  name: string, // The name of the field in schema
  value?: any, // The initial value. Coalesces to default value in schema
) => { ... };
```

### New function: `removeArrayFieldItem`

Should remove an array item in a given index for a given schema field

#### Identity

```jsx
const addArrayFieldItem = (
  name: string, // The name of the field in schema
  index: number, // The array index to remove
) => { ... };
```

**Comment by LauraBeatris** [comment]
How can we update programatically a field without relying on the ``onChange`` handler?

**Comment by gabrielnaoto** [comment]
What would happen if we passed a non `array` path from schema to the `FormFieldArray ` can formup validate and deal with it without breaking?

Example: throw a console error `Invalid path: path must be of type array`?

**Comment by gabrielnaoto** [comment]
> How can we update programatically a field without relying on the `onChange` handler?

I think, if the path can handle array indexes, we could do what we do right now, passing the full path, right?

**Comment by pedro-lb** [comment]
> What would happen if we passed a non `array` path from schema to the `FormFieldArray ` can formup validate and deal with it without breaking?
> 
> Example: throw a console error `Invalid path: path must be of type array`?

Exactly, I'll update the issue description to note that we should have this invariant.

**Comment by pedro-lb** [comment]
> How can we update programatically a field without relying on the `onChange` handler?

We could simply fallback to `formikForm.setFieldValue("arrayField", value);`

**Comment by gabrielpra1** [comment]
What's the reasoning behind the `FormFieldArray` name? Isn't it a `InputGroup`?

I also think that "array field" reads better than "field array". What do you think?

**Comment by pedro-lb** [comment]
> > The calculated path, so the user doesn't need to string interpolate.
> 
> User still needs to interpolate, though, right?
> 
> ```js
>             <FormInput
>               name={`${item.path}.name`} // here we're accessing "name" inside the array. Full path = people[index].name
>               // value={item.value.name} // we can also access the value at the index in "item"
>             />
> ```
> 
> Maybe we could provide a function to do that? To be called like `name={item.name("name")}`

We already have a component named `FormInputGroup`, which is used for groups of inputs that can be selected (e.g. an arrya of radio buttons or checkboxes), that's why I opted out for `group` in the name - it's better to leave `array` as explicit.

Other than that, the names are up for discussion! 

I agree that `FormArrayField` looks better than `FormFieldArray` 😃

**Comment by pedro-lb** [comment]
> > The calculated path, so the user doesn't need to string interpolate.
> 
> User still needs to interpolate, though, right?
> 
> ```js
>             <FormInput
>               name={`${item.path}.name`} // here we're accessing "name" inside the array. Full path = people[index].name
>               // value={item.value.name} // we can also access the value at the index in "item"
>             />
> ```
> 
> Maybe we could provide a function to do that? To be called like `name={item.name("name")}`

Looks really good!

However, for primitive types, we don't need to interpolate.

We could do:

- `item.name`: string that already interpolates the path (e.g. `phoneNumbers[3]`)
- `item.getName("...")`: function that interpolates the path (e.g. `people[3].email`)

**Comment by pedro-lb** [comment]
PR https://github.com/jungsoft/formup/pull/33

**Comment by pedro-lb** [comment]
# Implementation changes

- `addArrayFieldItem` and `removeArrayFieldItem` are not exported by `useFormup`, instead, they are passed as arguments & props by `FormArrayField`:

```jsx
<FormArrayField name="colors">
  {(items, arrayHelpers) => (
    // arrayHelpers.push() // adds a new default item
    // arrayHelpers.push('3') // adds a new valued item
    // arrayHelpers.remove(index) // removes an item
  )}
</FormArrayField>
```

# Known issues

- After removing an item in the array & adding another one, the `touched` state is kept as the old one:

![image](https://user-images.githubusercontent.com/22061051/87965590-6498e880-ca92-11ea-8c2b-83f3c9199e69.png)

This seems to be a bug in formik, and will be fixed in `2.0` according to roadmap.

**Comment by pedro-lb** [comment]
# Known issues

- After removing an item in the array & adding another one, the `touched` state is kept as the old one:

![image](https://user-images.githubusercontent.com/22061051/87965590-6498e880-ca92-11ea-8c2b-83f3c9199e69.png)

This seems to be a bug in formik, and will be fixed in `2.0` according to roadmap.

**Comment by pedro-lb** [comment]
Released in https://github.com/jungsoft/formup/releases/tag/1.5.0
